src/afa/afa.cc

```diff
@@ -34,29 +36,531 @@ const std::string Mata::Afa::TYPE_AFA = "AFA";
+        // The clause (1 && 2) could be deleted. Solution:
+        // create closed set -> add the new transition -> extract the antichain ->
+        // store it to the transition (???). Working solution:
+        //        
+        // auto cl = ClosedSet<State>(ClosedSet<State>::upward_closed, 0, max_state_used, *nodesPtr);
+        // cl.insert(trans.dst);
+        // *nodesPtr = cl.get_antichain();
+        //        
+        // Is it worth it?         
+
+        *nodesPtr = nodesPtr->Union(trans.dst);
+        return;
+    }
+    transRelation[trans.src].push_back(trans);
+
 } // add_trans }}}
 
+
+/** This function adds a new inverse transition to the automaton
+* It changes an inverse transition relation.
+* @brief adds a new inverse transition
+* @param trans a given TRANSITION (it will be inverted within this function)
+*/
+void Afa::add_inv_trans(const Trans& trans)
+{// {{{
+
+    // Iterating through all the nodes which were given as a destination of the current transition
+    for(auto node : trans.dst)
+    {
+        // TODO: Could this ever happen?
+        if(node.empty())
+        {
+            continue;
+        }
+
+        // If there is already a memory cell which corresponds to the current dst node and a transition symbol,
+        // we have to find it and add the 'src' state to the corresponding result_nodes vector. 
+        // Let us recall that the inverse transition datatype is a vector of vectors of tuples (symbol, vector_of_pointers)
+        // Each mentioned pointer points to the tuple (result_nodes, sharing_list). If there exists such a tuple, where
+        // the sharing_list corresponds to the current dst node, we can simply add the current 'src' state to the result_nodes.
+        // Since the corresponding tuple (result_nodes, sharing_list) must be referred from all the vector_of_pointers, where
+        // symbol == trans.symbol, it is sufficient to choose the first element of the dst node and look if there exists such a
+        // tuple (result_nodes, sharing_list), where sharing_list == dst node in context of the current symbol
+        //
+        // Example: We have transitions (0, a, {0, 1}), (0, b, {1}). The inverse transition data structure looks like this:
+        //
+        // 0 -> {('a', {&(result_nodes:{0}, sharing_list:{0, 1})})}
+        // 1 -> {('a', {&(result_nodes:{0}, sharing_list:{0, 1})}),
+        //      ('b', {&(result_nodes:{0}, sharing_list:{1})})}
+        //
+        // Now, we want to add the transition (1, a, {0, 1}), so we choose only one element of the dst node (0 or 1) and look if there exists
+        // a tuple (result_nodes, sharing_list), where sharing_list == {0, 1}. If so, we add 0 to the result_nodes. 
+        bool found = false;
+
+        auto invTransPtrs = perform_inv_trans(*(node.begin()), trans.symb);
+        for(auto result : invTransPtrs)
+        {
+            if(result->sharing_list == node)
```

Comment:
Looking at the use of `sharing_list` now, I really do not like the name. I have no idea from the name itself, what is it supposed to do. Could we maybe come with a more appropriate name?

Reply:
Thanks for your suggestion, we have discussed it. In the latest version of the code, I use the name `precondition`.